lib/SILOptimizer/IPO/GlobalOpt.cpp

```diff
@@ -770,6 +804,45 @@ void SILGlobalOpt::optimizeInitializer(SILFunction *AddrF,
   HasChanged = true;
 }
 
+/// If there are no loads or accesses of a given global, then remove its
+/// associated global addr and all asssociated instructions.
+bool SILGlobalOpt::tryRemoveGlobalAddr(SILGlobalVariable *global) {
+  if (isPossiblyUsedExternally(global->getLinkage(), Module->isWholeModule()))
+    return false;
+  if (GlobalLoadMap.count(global) || GlobalAccessMap.count(global))
+    return false;
+
+  for (auto *addr : GlobalAddrMap[global]) {
+    eraseUsesOfInstruction(addr);
+    addr->eraseFromParent();
+  }
+
+  reset();
```

Comment:
I worry that calling collect() every time something changes is going to turn this pass into an O(n^2) algorithm. Is there a way to build a list of all instructions to be deleted, and then delete them at the end, and finally reset()?

Reply:
Before deleting globals, I need to recollect everything in order to figure out what's now-dead. This will make it an O(n2) -- is that OK?

Reply:
That's OK (and it answers one of my questions in the latest review) but my guess would be that it should be possible to determine in one pass both which instructions to delete, and which SIL globals become dead as a result.

Reply:
Yes, see [my comment](https://github.com/apple/swift/pull/28324#discussion_r356409486).